### PR TITLE
Implement caching for dependency expansion

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This adds opt-in caching to `Dependency.expand` and `Requirement.expand`.

This massively increases the speed of those functions, if a cache key is passed. I've adjusted `Formula#rescursive_dependencies` and `Formula#recursive_requirements` to do this if no block is passed.

This massively improves the `brew uses` runtime from ~40s on Linux to about 6s, with most of that remaining time now being the actual `Formula.to_a` file loading.

Care is needed when choosing a cache key - the block is not re-evaulated on cache hits and so the cache key should be constructed depending on what variables might require rerunning the block.

I tried to be careful to clear the cache where it might be necessary - hopefully I haven't missed an edge case.

If successful, this should also speed up the `test default formula (Linux)` CI step.

### CODE REVIEW

File: Library/Homebrew/formula.rb [hunk 1]
```diff
@@ -222,10 +222,28 @@ def active_spec=(spec_sym)
...
+    Requirement.clear_cache
+  end
+
+  # @private
+  def build=(build_options)
+    old_options = @build
+    @build = build_options
+
+    return if old_options&.used_options == build_options&.used_options &&
+              old_options&.unused_options == build_options&.unused_options
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:243]: nil to nil comparisons here are fine I assume?

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense, nice work! The only thing I'd be tempted to do here is make the interface a bit more similar to the existing `uses` caching i.e. `Formulary.enable_factory_cache!` or something similar/shared which allows the cache keys to be set inside the relevant methods.

**Comment by Bo98** [comment]
> something similar/shared which allows the cache keys to be set inside the relevant methods.

The hard bit is that the return value of `Dependency.expand` depends on the block passed, which varies from caller to caller.

File: Library/Homebrew/formula.rb (hunk 1 above)
Comment (Bo98) [reply Library/Homebrew/formula.rb:243]: I'll relook into this. I'm not really convinced this should ever be nil - it just was in some test.

Comment (Bo98) [reply Library/Homebrew/formula.rb:243]: https://github.com/Homebrew/brew/blob/e143dfdad6bbd4cd178906eccac5de095c80dc2f/Library/Homebrew/test/formula_spec.rb#L168-L172

Not really convinced `nil` is something we should support. I don't see any `build` nil checks anywhere.

**Comment by Bo98** [comment]
> If successful, this should also speed up the `test default formula (Linux)` CI step.

About 12 minutes -> less than 2 minutes

**Comment by MikeMcQuaid** [comment]
> About 12 minutes -> less than 2 minutes

🤯 

> The hard bit is that the return value of `Dependency.expand` depends on the block passed, which varies from caller to caller.

Ah, I see. No worries, then.

**Review by MikeMcQuaid — APPROVED** [review]
